x/leverage/README.md

```diff
@@ -154,52 +175,146 @@ Supply utilization of a token denomination is calculated as:
 
-- Borrowed value must be less than collateral value times `CollateralWeight` (sum over each collateral asset)
-- Borrowed value times `BorrowFactor` (sum over each borrowed asset) must be less than collateral value.
+#### Special Asset Pairs
 
-This means that when the original borrow limit based on collateral weight would allow a higher quality collateral to borrow a risky asset with a small margin of safety, the user's effective collateral weight is reduced to that of the riskier asset.
-(Or `0.5` at the minimum.)
+The leverage module can define pairs of assets which are advantaged when one is used as collateral to borrow the other.
 
-#### Historic Borrow Limit, Value
+They are defined in the form `[Asset A, Asset B, Special Collateral Weight]`. In effect, this means that
 
-The leverage module also makes use of the oracle's historic prices to enforce an additional restriction on borrowing.
+> When a user has collateral of `Asset A` and borrows `Asset B`, or vice versa, the `CollateralWeight` of both `Asset A` and `Asset B` are replaced by `Special Collateral Weight`.
 
-The logic is:
+#### Borrow Limit
 
-- For any `MsgBorrow`, `MsgMaxBorrow`, `MsgDecollateralize`, `MsgWithdraw`, or `MsgMaxWithdraw`
-- The borrower’s borrowed value must be less than their borrow limit, with borrowed value being computed using `PriceModeHigh`, i.e. the higher of either spot price or historic price is used.
-- Where historic prices are defined as the Median of the last `N` historic medians from the `oracle` module with `N = Token.HistoricMedians` in the leverage registry
-- Else the transaction fails
+A user's borrow limit is the sum of the contributions from each denomination of collateral they have deposited, with some modifications due to `Borrow Factor` and `Special Asset Pairs`.
+
+The full calculation of a user's borrow limit is as follows:
+
+1. Calculate the USD value of the user's collateral assets, using the _lower_ of either spot price or historic price for each asset. Collateral with missing prices is treated as zero-valued.
+2. Calculate the USD value of the user's borrowed assets, using the _higher_ of either spot price or historic price for each asset. Borrowed assets with missing prices cause any transaction which could increased borrowed value or decrease borrow limit to fail.
+3. Sort all `Special Asset Pairs` with assets matching parts of the user's position, starting with the highest `Special Collateral Weight`.
+4. For each special asser pair, match collateral tokens with borrowed tokens until one of the two runs out. The matched amounts satisfy `Collateral Value (A) * Special Collateral Weight (A,B) = Borrowed Value (B)` for each special asset pair `[A,B,CW]`. Subtract the collateral and borrowed tokens from the user's remaining position.
+5. Then sort the user's remaining collateral tokens by `Collateral Weight` and sort their remaining borrowed tokens by  the same.
```

Comment:
The current implementation in draft is resilient to this case - it will use the asset's collateral weight if the special pair would decrease efficiency